node_manager.go

```diff
@@ -0,0 +1,965 @@
+	var node Node
+
+	// Grab the current context.
+	ctx, cancel := context.WithTimeout(context.Background(),
+		lcfg.EtcdGracePeriodSec * time.Second)
+
+	// Get a response from the primed node.
+	response, err := inst.internal.Get(ctx, path.Join(lcfg.Nodes_dir, hid),
+		clientv3.WithSort(clientv3.SortByKey, clientv3.SortAscend))
+
+	// if debug mode...
+	if lcfg.DebugMode && err == nil {
+
+		// cycle thru all of the current nodes for the benefit of the developer
+		debugf("The primed node contains the following:")
+		for _, ent := range response.Kvs {
+			debugf(string(ent.Key) + " => " + string(ent.Value))
+		}
+	}
+
+	// cancel the current context as it is no longer needed
+	cancel()
+
+	// if an error occurs, pass it back
+	if err != nil {
+		return err
+	}
+
+	// further safety check, ensure the primed node is not null
+	if len(response.Kvs) < 1 {
+		return errorf("QueueJobOnNode() --> primed node appears to be null")
+	}
+
+	// unmarshal the recovered response data
+	err = json.Unmarshal(response.Kvs[0].Value, &node)
+
+	// if an error occurs, pass it back
+	if err != nil {
+		return err
+	}
+
+	// Job ID is the next index in the array.
+	uuid := strconv.FormatInt(j.Pid, 10)
+
+	// Marshal the job data.
+	mresult, err := json.Marshal(j)
+
+	// if an error occurred, pass it back
+	if err != nil {
+		return err
+	}
+
+	// Since the primed node gave back a valid response, this needs to
+	// place the job on the queue of the node; so firstly assemble the
+	// queue path on etcd as per the following:
+	//
+	// key => /nodes/jobs/uuid
+	//
+	// Developer note: the queued job of uuid is waiting, and its details
+	// can be found in the global 'processesList'.
+	//
+	jobQueue := path.Join(lcfg.Nodes_dir, hid, "jobs")
+
+	// grab the current context, need it to hand off the job to the node
+	ctx, cancel = context.WithTimeout(context.Background(),
+		lcfg.EtcdGracePeriodSec * time.Second)
+
+	// Insert the job into the job queue of the node.
+	responseToJobAddToNode, err := inst.internal.Put(ctx,
+		path.Join(jobQueue, uuid), string(mresult))
+
+	// if debug mode...
+	if lcfg.DebugMode && err == nil && responseToJobAddToNode != nil {
+
+		// tell the developer what happened at this stage...
+		debugf("Node manager has responsed with valid put response.")
+	}
+
+	// cancel the current context as this no longer needs it
+	cancel()
+
+	// if an error occurs, pass it back
+	if err != nil {
+		return err
+	}
+
+	// update etcd with the new node queue info
+	err = inst.putNode(&node)
+
+	// if an error occurred, pass it back
+	if err != nil {
+		stdlog(err.Error())
+		return err
+	}
+
+	// everything was a success, so return nil here
+	return nil
+}
+
+//! Grab the node of a given host.
+/*
+ * @param    string    host id
+ *
+ * @return   Node*     pointer to a given node
+ * @return   error     error message, if any
+ */
+func (inst *EtcdInstance) getNode(hid string) (*Node, error) {
```

Comment:
Yes, hid is far too generic. I have since changed it.

https://github.com/rbisewski/lclusterd/commit/a12b1701db557077c4841872609c2a2d13391fa4